File: backend/src/AnimeShop/API/Maintenance/EquipmentInspectionResults.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Maintenance.EquipmentInspectionResults where

import AnimeShop.Prelude

type EquipmentInspectionResultsAPI =
  "equipment_inspection_results" :>
  (
    Get '[JSON] [Entity EquipmentInspectionResult] :<|>
    ReqBody '[JSON] EquipmentInspectionResult :> Post '[JSON] (Entity EquipmentInspectionResult) :<|>
    ReqBody '[JSON] EquipmentInspectionResultId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity EquipmentInspectionResult) :> Patch '[JSON] ()
  )

equipmentInspectionResultsServer :: ServerT EquipmentInspectionResultsAPI App
equipmentInspectionResultsServer = getEquipmentInspectionResults :<|> postEquipmentInspectionResult :<|> deleteEquipmentInspectionResult :<|> patchEquipmentInspectionResult

getEquipmentInspectionResults :: App [Entity EquipmentInspectionResult]
getEquipmentInspectionResults = runDb $ selectList [] []

postEquipmentInspectionResult :: EquipmentInspectionResult -> App (Entity EquipmentInspectionResult)
postEquipmentInspectionResult = runDb . insertEntity

deleteEquipmentInspectionResult :: EquipmentInspectionResultId -> App ()
deleteEquipmentInspectionResult = runDb . delete

patchEquipmentInspectionResult :: Entity EquipmentInspectionResult -> App ()
patchEquipmentInspectionResult entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/maintenance/equipmentInspectionResults.js
import { ADD_EQUIPMENT_INSPECTION_RESULT, GET_EQUIPMENT_INSPECTION_RESULTS, DELETE_EQUIPMENT_INSPECTION_RESULT } from '../../actions/types';

export function equipmentInspectionResults(state = [], action) {
  switch (action.type) {
    case ADD_EQUIPMENT_INSPECTION_RESULT:
      return [ ...state, action.equipmentInspectionResult];
    case GET_EQUIPMENT_INSPECTION_RESULTS:
      return action.equipmentInspectionResults;
    case DELETE_EQUIPMENT_INSPECTION_RESULT:
      return state.filter(equipmentInspectionResult => equipmentInspectionResult.id !== action.id);
    default:
      return state;
  }
}
